archinstall/lib/user_interaction.py

```diff
@@ -939,3 +939,86 @@ def select_locale_enc(default):
 	).run()
 
 	return selected_locale
+
+def generic_select(p_options :Union[list,dict],
+				input_text :str = "Select one of the above by index or absolute value: ",
+				allow_empty_input :bool = True,
+				options_output :bool = True,   # function not available
+				sort :bool = False,
+				multi :bool = False,
+				default :Any = None) -> Any:
+	from copy import copy
+	"""
+	A generic select function that does not output anything
+	other than the options and their indexes. As an example:
+
+	generic_select(["first", "second", "third option"])
+		first
+		second
+		third option
+	When the user has entered the option correctly,
+	this function returns an item from list, a string, or None
+
+	Default value must be in the list of options
+	"""
+	# I (shallow) copy the options object. This is an input only parameter
+	options = copy(p_options)
+	# Checking if the options are different from `list` or `dict` or if they are empty
+	if type(options) not in [list, dict]:
```

Comment:
I wonder if it wouldn't be better to move those validations of menu parameters into the menu class since it's the same for every call essentially

Reply:
Haven't checked if they already are. I lifted them directly from the old version (with the exception of the copy, which is my particular obsession).
Will have a look

Reply:
```python
	# I (shallow) copy the options object. This is an input only parameter
	options = copy(p_options)
```
I need it very much propagated (see below about sort), but it needs to stay here (I do some specific preprocessing)
 ```python
	# Checking if the options are different from `list` or `dict` or if they are empty
	if type(options) not in [list, dict]:
		log(f" * Generic select doesn't support ({type(options)}) as type of options * ", fg='red')
		log(" * If problem persists, please create an issue on https://github.com/archlinux/archinstall/issues * ", fg='yellow')
		raise RequirementError("generic_select() requires list or dictionary as options.")
	if not options:
		log(" * Generic select didn't find any options to choose from * ", fg='red')
		log(" * If problem persists, please create an issue on https://github.com/archlinux/archinstall/issues * ", fg='yellow')
		raise RequirementError('generic_select() requires at least one option to proceed.')
```
both can be moved to menu, but add some new dependencies (log and the exception). What do you think ?
```python
	# After passing the checks, function continues to work
	if type(options) == dict:
		# To allow only `list` and `dict`, converting values of options here.
		# Therefore, now we can only provide the dictionary itself
		options = list(options.values())
```
This is exactly the opposite of what Menu does (it uses the keys as entries by default). Must stay here
```python
	# check that the default value is in the list
	if default and default not in options:
		log(f" * Default option {default} not in the list of options ", fg='red')
		log(" * If problem persists, please create an issue on https://github.com/archlinux/archinstall/issues * ", fg='yellow'
		raise RequirementError('generic_select() requires the default to be on the list.')
```
AFAIK Menu does not make any assumption about it, which i'd liked to do. But it helps me to make a cleaner code ;-)
Don't think it should be propagated, in general. Even here, it could eventually disappear
```python 
	if sort:
		# As we pass only list and dict (converted to list), we can skip converting to list
		options = sorted(options)
```
Ooops, the whole sort can create very difficult to trace bugs, unless inmutability of the lists can be guaranteed, and even then as the sort here and the one at menu are not against the same object type.

The next upload will have the copy and sort questions "sorted out". I wait to your feedback about pushing depenedencies into menu